### PR TITLE
REST high-level client: add validate query API

### PR DESCRIPTION
**Pull request by sohaibiftikhar**
Relates to #27205

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java
Comment (sohaibiftikhar) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java]: @nik9000 @javanna Since its hard to compare exceptions, in this case, I just check for the index and the shardId. Is this alright?

File: server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java
Comment (sohaibiftikhar) [inline server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java]: @nik9000 @javanna I also added this for the same reason. Since exception messages change after deserialization with `ElasticSearchException`.

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: I'd use `Objects.hash` because the performance here isn't that important and it is easier to review for correctness.

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: Can you make this look like our other equals methods?

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java]: I'd prefer to keep the `equals` method complete and only use it in the test when there aren't exceptions. I dunno what we've done other places though.

I'd also like to do something like `assertThat(exception.getMessage(), containsString("original message"));`.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java]: :+1:

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by nik9000** [comment]
@elasticmachine, whitelist this PR please.

**Comment by nik9000** [comment]
@elasticmachine, add to whitelist

File: test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java
Comment (sohaibiftikhar) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java]: @nik9000 So what do you reckon? I shouldn't extend `AbstractStreamableXContentTestCase`?

Comment (nik9000) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java]: I think you can extend that to get good randomized coverage for the case when you don't have any exceptions and then hand write a couple of test cases for the exception round tripping. I *think* @javanna talked about doing that some other place....

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (sohaibiftikhar) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: Thanks! I didn't know this existed either :)

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java
Comment (sohaibiftikhar) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java]: The assert part makes sense. However, I don't completely understand what needs to be done for the `equals`.

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java]: I mean to make `equals` just call `equals` on the Exceptions. It can certainly do `Objects.equals` for the `null`, but I'd prefer the equals make sure the objects are actually equal. And because we can't assert proper equality when round tripping the exceptions I'd hand write an test for the round tripping with exceptions and leave exceptions out of the `AbstractStreamableTestCase` infrastructure.

**Comment by sohaibiftikhar** [comment]
I was not able to reproduce the errors on my machine. Will merge in master to see if that fixes it. There is a remote possibility that these are related though.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: Can you add a test to https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java

File: docs/java-rest/high-level/indices/validate_query.asciidoc
Comment (hub-cap) [inline docs/java-rest/high-level/indices/validate_query.asciidoc]: `ValidateQueryRequest`

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: it looks like we might be missing a few more options such as `types` and `indicesOptions`, as specified in https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestValidateQueryAction.java#L62-L67

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (sohaibiftikhar) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: Yes, thanks for catching this.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (sohaibiftikhar) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: `types` are being set in the URL I believe. I will check the `indicesOptions`.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java
Comment (javanna) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java]: Have a look at `ListTasksResponseTests`, where we separate testing between two scenarios: without (no big problems, assertToXContentEquivalence true) and with (assertToXContentEquivalence false). I think that this change may not be needed if you do it that way.

**Comment by sohaibiftikhar** [comment]
@nik9000 Could you have a look again when you have time?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: can you please add params tag like we have now in all of the other methods?

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: nit: can you remove some of these empty lines please?

File: docs/java-rest/high-level/indices/validate_query.asciidoc
Comment (javanna) [inline docs/java-rest/high-level/indices/validate_query.asciidoc]: can you add an example of how to set the query to the request?

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: is there a reason not to use Objects.equals like we do in a lot of other equals methods?

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java]: can you make it a ToXContentObject please instead?

File: server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java]: why do we need to shuffle these two?

File: server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java]: do these methods need to be protected?

**Review by javanna — CHANGES_REQUESTED** [review]
left a couple of comments as well.

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (sohaibiftikhar) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: You are right. I will switch this.

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java
Comment (sohaibiftikhar) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java]: In this case, I should add `builder.startObject` in the `toXContent` method, right?

File: server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java
Comment (sohaibiftikhar) [reply server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java]: Umm no I think. I will make it `private`.

File: server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java
Comment (sohaibiftikhar) [reply server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java]: We are creating the XContent thing very serially. First all successful ones and then all failures. I thought it would be more realistic if this was shuffled.

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java]: yea startObject and endObject, see #31155 also

File: server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java
Comment (javanna) [reply server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java]: but they go to two different lists, so I think that what matters is that each item is randomized? I am not sure but shuffling buys us.

Comment (sohaibiftikhar) [reply server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java]: The query explanation contains the failures and the successful ones. Ok, we can do away with the shuffling of failures. But the query explanations should stay I think.

File: docs/java-rest/high-level/supported-apis.asciidoc
Comment (javanna) [inline docs/java-rest/high-level/supported-apis.asciidoc]: this may be why the docs don't build? :)

Comment (sohaibiftikhar) [reply docs/java-rest/high-level/supported-apis.asciidoc]: Yes. Thanks. Seems I forgot to build the docs after merging...

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (nik9000) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: Could you revert these formatting changes that aren't part the change? It make git blame harder to work with and I kind of like the old formatting better anyway.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java
Comment (nik9000) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: This formatting change is fine though, because the old code was just wrong.

Comment (nik9000) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: I'm aware that this is somewhat contradictory.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java
Comment (nik9000) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]: I think this change is fine but I'd prefer it in a separate PR.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java
Comment (nik9000) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]: We want to keep the `"`s aligned.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java
Comment (nik9000) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]: Can you revert the formatting change? I think the old way is better. If we want to do it we should do it in a separate PR anyway.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java
Comment (nik9000) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]: Can you revert all of these formatting changes above?

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: I tend to use the members rather than the getters for these but either way is ok by me!

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java]: I think this'd be better formatted like:
```
builder.startObject();
builder.field("query");
query.toXContent(builder, params);
return builder.endObject();
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java
Comment (sohaibiftikhar) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]: I think I messed up something during the merge from master and my editor did all these formats. I will revert everything back.

File: server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java
Comment (sohaibiftikhar) [reply server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/QueryExplanation.java]: I used to prefer that too. Then I saw in some classes that the getters also had some logic which I wanted to use. So I started using these...

**Review by nik9000 — APPROVED** [review]
LGTM.  @hub-cap, would you like another look?

**Comment by nik9000** [comment]
> LGTM. @hub-cap, would you like another look?

Woops, not @hub-cap, @javanna.

**Review by javanna — APPROVED** [review]
LGTM thanks @sohaibiftikhar

**Comment by camerondavison** [comment]
Is there a reason that some methods in this api such as https://github.com/elastic/elasticsearch/blob/7a6021ca98d3661eb6d8e80bda033b1b6d8c2d9d/server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryRequest.java#L114 do not return this `ValidateQueryRequest`

**Comment by javanna** [comment]
hi @camerondavison the reason is merely that newly added methods do not return the request, while old methods do. We would prefer to have pure setters in our codebase at this point, but we don't see the value in such a big breaking change, hence we only apply the preferred standard when adding new methods for now.

**Pull request merged**
